src/compiler/program.ts

```diff
@@ -194,7 +197,117 @@ namespace ts {
             getEnvironmentVariable: name => system.getEnvironmentVariable ? system.getEnvironmentVariable(name) : "",
             getDirectories: (path: string) => system.getDirectories(path),
             realpath,
-            readDirectory: (path, extensions, include, exclude, depth) => system.readDirectory(path, extensions, include, exclude, depth)
+            readDirectory: (path, extensions, include, exclude, depth) => system.readDirectory(path, extensions, include, exclude, depth),
+            createDirectory: d => system.createDirectory(d)
+        };
+        return compilerHost;
+    }
+
+    /*@internal*/
+    export function changeCompilerHostToUseCache(
+        host: CompilerHost,
+        toPath: (fileName: string) => Path,
+        useCacheForSourceFile: boolean
+    ) {
+        const originalReadFile = host.readFile;

```

Comment:
In general it's not usually safe to pluck methods off a host like this. If an API uses passes in a class instance for a host, this can unbind the `this` for these methods. Best `.bind` these to `host` here so that doesn't happen.

Reply:
We are always calling it by passing host as this to these functions.calls. For now these are passed only to reset back host and once we move to new api we wont need to return these at all  and hence its safe to not bind it.